### PR TITLE
Assemble `Unpin` candidates specially for generators in new solver

### LINKED ISSUE #16: Non-static generators should always be unpin
```
#![feature(generators, generator_trait)]

use std::marker::{PhantomPinned, Unpin};

fn assert_unpin<G: Unpin>(_: G) {
}

fn main() {
    // Even though this generator holds a `PhantomPinned` in its environment, it
    // remains `Unpin`.
    assert_unpin(|| {
        let pinned = PhantomPinned;
        yield;
        drop(pinned);
    });
}
```

Just needs special logic added to builtin impls.

### PR DESCRIPTION
**Pull request by compiler-errors**
Fixes compiler-errors/next-solver-hir-issues#16

r? @lcnr

### CODE REVIEW

File: compiler/rustc_trait_selection/src/solve/trait_goals.rs [hunk 1]
```diff
@@ -168,6 +168,23 @@ impl<'tcx> assembly::GoalKind<'tcx> for TraitPredicate<'tcx> {
 
             ty::Infer(_) | ty::Bound(_, _) => bug!("unexpected type `{self_ty}`"),
 
+            // Generators have one special built-in candidate, `Unpin`, which
+            // takes precedence over the structural auto trait candidate being
+            // assembled.
```
Comment (lcnr) [inline compiler/rustc_trait_selection/src/solve/trait_goals.rs:173]: conceptually what's happening here is that we have an *explicit* builtin impl of unpin for generators. the idea for auto traits is

```
fn consider_auto_trait_candidate(..) {
    if outmost_type_constructor_has_explicit_impl() {
        return Err(NoSolution);
    }

    add_auto_trait_candidate
}

fn consider_builtin_unpin_candidate(..) {
    // adds the candidate for generators.
}
```

where `outmost_type_constructor_has_explicit_impl` considers both user written and builtin impls. I am a bit worried that going forward we may try to add a builtin candidate for yet another auto trait but do so outside of `consider_auto_trait_candidate`. 

I don't think it really makes sense to actually have a `outmost_type_constructor_has_explicit_impl` and `consider_builtin_unpin_candidate` split. What do you think about

- move this whole match into a separate function, something like `disable_auto_trait_candidate_due_to_explicit_impl` which returns `Option<QueryResult<'tcx>>` with a comment mentioning that this function considers builtin candidates for auto traits
- add a comment in `assemble_builtin_impl_candidates` that for auto traits, builtin impls have to be added to `disable_auto_trait_candidate_due_to_explicit_impl` instead

Comment (compiler-errors) [reply compiler/rustc_trait_selection/src/solve/trait_goals.rs]: Hmm.. took a stab at that. Let me know what you think.

File: compiler/rustc_trait_selection/src/solve/trait_goals.rs [hunk 2]
```diff
@@ -618,6 +543,95 @@ impl<'tcx> assembly::GoalKind<'tcx> for TraitPredicate<'tcx> {
...
+            | ty::FnPtr(_)
+            | ty::Closure(_, _)
+            | ty::Generator(_, _, _)
+            | ty::GeneratorWitness(_)
+            | ty::GeneratorWitnessMIR(_, _)
+            | ty::Never
+            | ty::Tuple(_)
+            | ty::Error(_)
+            | ty::Adt(_, _)
+            | ty::Alias(ty::Opaque, _) => {
```
Comment (lcnr) [inline compiler/rustc_trait_selection/src/solve/trait_goals.rs:617]: `ty::Error` and opaque types feel weird here. Error shouldn't matter, but I am worried that this is wrong for opaques :thinking: Also unhappy that we treat opaques as `PlaceholderSimplifiedType` in `simplify_type`. Want to think more about this, can you add a fixme here?

apart from that r=me

**Comment by compiler-errors** [comment]
Added a FIXME for opaques and moved `ty::Error` to its own branch.

@bors r=lcnr rollup (only affects new solver)

**Comment by compiler-errors** [comment]
Unpin note got changed slightly

@bors r=lcnr

**Pull request merged**